### PR TITLE
refactor: rework aged authentication config interface

### PR DESCRIPTION
**Pull request by comtalyst**
Fixes # 

**Description**
Part of #224 
This involves refactoring of auth config, removing some redundant code (e.g., in favor of `DefaultAzureCredentials`, remove some unused variables), and renaming `ARM_USER_ASSIGNED_IDENTITY_ID` to `KUBELET_IDENTITY_CLIENT_ID` for accuracy to its purpose/use. See core motivations in the issue.

**How was this change tested?**
* Unit tests
* E2E - Utilization (through `make az-e2etests`)
* E2E for NAP (using managed identity)

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [X] No

**Release Note**

```release-note
- Deprecate `ARM_USE_CREDENTIAL_FROM_ENVIRONMENT` and `ARM_USE_MANAGED_IDENTITY_EXTENSION` in favor of track2-based authentication: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#defaultazurecredential
- Rename `ARM_USER_ASSIGNED_IDENTITY_ID` to `KUBELET_IDENTITY_CLIENT_ID` for accuracy to its purpose/use
- Remove unused environment variables `AZURE_CLUSTER_NAME` and `ARM_VM_TYPE`
```

### CODE REVIEW

**Review by comtalyst — COMMENTED** [review]
/test

File: pkg/auth/config.go [hunk 1]
```diff
@@ -165,22 +101,35 @@ func (cfg *Config) GetAzureClientConfig(authorizer autorest.Authorizer, env *azu
...
+	cfg.ResourceGroup = strings.TrimSpace(os.Getenv("ARM_RESOURCE_GROUP"))
+	cfg.TenantID = strings.TrimSpace(os.Getenv("ARM_TENANT_ID"))
+	cfg.SubscriptionID = strings.TrimSpace(os.Getenv("ARM_SUBSCRIPTION_ID"))
+	cfg.VMType = strings.ToLower(os.Getenv("ARM_VM_TYPE"))
+	cfg.ClusterName = strings.TrimSpace(os.Getenv("AZURE_CLUSTER_NAME"))
+	cfg.NodeResourceGroup = strings.TrimSpace(os.Getenv("AZURE_NODE_RESOURCE_GROUP"))
+	cfg.AuthMethod = strings.TrimSpace(os.Getenv("ARM_AUTH_METHOD"))
+	cfg.KubeletIdentityClientID = strings.TrimSpace(os.Getenv("ARM_KUBELET_IDENTITY_CLIENT_ID"))
+
+	return nil
```
Comment (Bryce-Soghigian) [inline pkg/auth/config.go:116]: if we are refactoring this right now can we also refactor them into the options struct?

Comment (comtalyst) [reply pkg/auth/config.go:116]: (discussed offline) that is being considered in the next PR since it would requires more refactor apart from the interface.

**Review by comtalyst — COMMENTED** [review]
/test

**Review by comtalyst — COMMENTED** [review]
/test

File: pkg/auth/cred.go [hunk 2]
```diff
@@ -30,26 +30,19 @@ func NewCredential(cfg *Config) (azcore.TokenCredential, error) {
 		return nil, fmt.Errorf("failed to create credential, nil config provided")
 	}
 
-	if cfg.UseCredentialFromEnvironment {
+	if cfg.ArmAuthMethod == authMethodWorkloadIdentity {
 		klog.V(2).Infoln("cred: using workload identity for new credential")
 		return azidentity.NewDefaultAzureCredential(nil)
```
Comment (tallaxes) [inline pkg/auth/cred.go:35]: "using workload identity" - this is not quite what `NewDefaultAzureCredential` does, it will [attempt a variaty of mechanisms](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#defaultazurecredential). If we are Ok with this - which may well be the case - let's just say it, and we don't need to special case Managed Identity below. If we want to constrain this to Workload Identity - needs to use `NewWorkloadIdentityCredential`?

**Review by tallaxes — COMMENTED** [review]
Before commenting on the appropriate configuration surface, and corresponding implementation, I would like to see a clear description of which authentication methods we want to support - and if we have a good reason to constrain them. Is there anything that prevents us from just using [`DefaultAzureCredential`](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#defaultazurecredential), which attempts a variety of authentication mechanisms (including the ones we need to support), using its existing/documented configuration surface? Would be good to avoid custom switches (like `ARM_AUTH_METHOD`) if the "standard"/OOB works well.

File: pkg/auth/cred.go (hunk 2 above)
Comment (comtalyst) [reply pkg/auth/cred.go:35]: There is a discussion related to this with @charliedmcb in another PR. I think our understanding match yours, with the new knowledge here is that `NewDefaultAzureCredential()` might cover managed identity as well.

So far, I think that we should scope it down to workload identity for simplicity. This means we will support only two auth methods for now: workload identity and system assigned MSI.
In the current implementation, there might be "shortcuts" where `ARM_AUTH_METHOD` is workload identity, but they ended up using something else that's compatible with this implementation. 

So, I think I will make changes to `NewWorkloadIdentityCredential`, if possible.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: I'm not sure what you mean by below; could you provide more example? 
>  Would be good to avoid custom switches (like ARM_AUTH_METHOD) if the "standard"/OOB works well.

Comment (tallaxes) [reply pkg/auth/cred.go:35]: What I mean is that `azidentity` already provides (via environment variables) an existing configuration surface, and (via `DefaultAzureCredential`) a reasonable sequence of auth methods to try, including the two that we need to support. We should have good reasons to a) constrain the auth methods, and b) introduce additional/custom configuration surface, such as the proposed `ARM_AUTH_METHOD` setting. (Which we may need to remove later if decide to use `NewDefaultAzureCredential`? Removal is always hard ...) There may be valid reasons for both (improved security posture, the built-in sequence not meeting our order requirements, established pattern elsewhere, etc.) - but would be good to have clarity on which ones - if any - apply here.

Comment (tallaxes) [reply pkg/auth/cred.go:35]: One more option to consider - if we decide there is a good reason to support only workload identity and MSI - is to just try them in order (similar to how `DefaultAzureCredential` does it), eliminating the need (and capability) for the explicit choice via `ARM_AUTH_METHOD`.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: >  b) introduce additional/custom configuration surface, such as the proposed ARM_AUTH_METHOD setting

`ARM_AUTH_METHOD` is in fact a replacement for the awkward combination of `ARM_USE_CREDENTIAL_FROM_ENVIRONMENT` and `ARM_USE_MANAGED_IDENTITY_EXTENSION` rather than an addition. 
But, if we use `DefaultAzureCredential`, maybe that can be implied from the credentials and the user don't have to explicitly specify `ARM_AUTH_METHOD`, eliminating its needs.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: > a) constrain the auth methods

I find that we never(?) have an expectation of the user using other auth methods than these two in the first place. Even if we allow it, that allow means additional responsibility/accountability (outside of this relatively simple implementation, I mean) on our side. 
So, unless there are good demands for other auth methods that is possible through `DefaultAzureCredential`, I don't think we should give the opportunity for the user to mess up things that we not yet want to invest into.

But at least, our code framework should be able to support that quickly once we decided to, but not exposing it to the user.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: > Removal is always hard ...
Removing ARM_AUTH_METHOD to eliminate user-side complexity, given `DefaultAzureCredential` can retrieve that implicitly, is an annoyance on our side to change the code, but I think it is ok.

On the other hand, if the user secretly rely on some auth methods that `DefaultAzureCredential` supports, but then we decide to remove it (e.g., security), then that's going to affect them.
Technically, if we stop using it and switch to `NewWorkloadIdentityCredential`, it is like that. But given we are in the early stage, I think it is better to do it now.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: Keep in mind that whatever in `DefaultAzureCredential` today might changes in the future under our radar, given its definition is open to new methods rather than always specific to a few already.

Comment (tallaxes) [reply pkg/auth/cred.go:35]: My preference would be for the simplest secure approach that supports the requirements, which seems to be `DefaultAzureCredential`. (Any changes there will likely require additional configuration, and are very unlikely to break existing auth scenarios. In NAP we control the configuration tightly enough for this not to be a concern, for self-hosted breaking changes, if any, would/should surface via release notes.) Second best would be to try a more controlled sequence of auth methods we need to support - unless you see a problem with this, and we have to specify the method explicitly via configuration.

Comment (Bryce-Soghigian) [reply pkg/auth/cred.go:35]: Agreed DefaultAzureCredential already encapsulates what we need. I don't see them breaking it either as its wildly used in many projects for wrapping many types of auth based on some quick github code searching.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: > Any changes there will likely require additional configuration, and are very unlikely to break existing auth scenarios

> I don't see them breaking it either

I doubt it will break too, plus the tests will catch it anyway in case we neglect release notes. 
But my point is it might have new auth methods we are not aware of, which means more liability as I mentioned. This will not happen in the more controlled approach (i.e., `NewWorkloadIdentityCredential` with `ARM_AUTH_METHOD`)

>  In NAP we control the configuration tightly enough for this not to be a concern

Yep all these concerns are for self-hosted. In NAP we have static integration interface.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: But maybe I am overthinking again. We could try going with implicit auth method and use `DefaultAzureCredential` and see how it goes.

The only problem, if it is at all, is `autorest_auth.go` is also using it. But I just haven't take a look at that deeply yet. If we are moving in this direction then I will see the possibility.

Comment (comtalyst) [reply pkg/auth/cred.go:35]: Circling back to this. 
- The concern of unwanted auth method is valid, from an example in Cluster Autoscaler. "Requirement" to continue the support is an engineering burden.
- Although, this `DefaultAzureCredential` is something standardized. If we are going to rely on this standardization (cloud-provider-azure?) for a next while, the simplicity we get might worth it, and the deprecation (if necessary) might not be too difficult.

TL;DR let's just use `DefaultAzureCredential`.

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -169,7 +170,7 @@ func getCABundle(restConfig *rest.Config) (*string, error) {
 }
 
 func getVnetGUID(cfg *auth.Config, subnetID string) (string, error) {
-	creds, err := auth.NewCredential(cfg)
+	creds, err := azidentity.NewDefaultAzureCredential(nil)
```
Comment (comtalyst) [inline pkg/operator/operator.go:173]: An alternative is to keep the abstraction, but just return `azidentity.NewDefaultAzureCredential(nil)`. But there are only 2 uses (another in azure_client.go) at the moment, so I don't think it is worth it.

File: pkg/providers/instance/skuclient/skuclient.go [hunk 4]
```diff
@@ -48,11 +50,14 @@ func (sc *skuClient) updateInstance() {
...
-	authorizer, err := auth.NewAuthorizer(sc.cfg, sc.env)
+	// Create a new authorizer for the sku client
+	// TODO (charliedmcb): need to get track 2 support for the skewer API
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
 	if err != nil {
-		klog.V(5).Infof("Error creating authorizer for sku client: %s", err)
+		klog.V(5).Infof("Error creating authorizer for sku client: default cred: %s", err)
 		return
 	}
+	authorizer := azidext.NewTokenCredentialAdapter(cred, []string{azidext.DefaultManagementScope})
```
Comment (comtalyst) [inline pkg/providers/instance/skuclient/skuclient.go:60]: An alternative is to keep the abstraction, that returns authorizer. But this is the only use of it at the moment, so I don't think it is worth it. Could refactor later.

**Review by comtalyst — COMMENTED** [review]
/test

**Review by comtalyst — COMMENTED** [review]
/test

File: pkg/auth/config.go [hunk 5]
```diff
@@ -165,22 +90,30 @@ func (cfg *Config) GetAzureClientConfig(authorizer autorest.Authorizer, env *azu
...
-	cfg.NodeResourceGroup = strings.TrimSpace(cfg.NodeResourceGroup)
+func (cfg *Config) Build() error {
+	// May require more than this behind the scenes: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#defaultazurecredential
+	cfg.Cloud = strings.TrimSpace(os.Getenv("ARM_CLOUD"))
+	cfg.Location = strings.TrimSpace(os.Getenv("LOCATION"))
+	cfg.ResourceGroup = strings.TrimSpace(os.Getenv("ARM_RESOURCE_GROUP"))
+	cfg.TenantID = strings.TrimSpace(os.Getenv("ARM_TENANT_ID"))
+	cfg.SubscriptionID = strings.TrimSpace(os.Getenv("ARM_SUBSCRIPTION_ID"))
+	cfg.VMType = strings.ToLower(os.Getenv("ARM_VM_TYPE"))
+	cfg.ClusterName = strings.TrimSpace(os.Getenv("AZURE_CLUSTER_NAME"))
```
Comment (Bryce-Soghigian) [inline pkg/auth/config.go:101]: How is this used?

File: pkg/providers/instance/azure_client.go [hunk 6]
```diff
@@ -108,11 +109,11 @@ func CreateAZClient(ctx context.Context, cfg *auth.Config) (*AZClient, error) {
 }
 
 func NewAZClient(ctx context.Context, cfg *auth.Config, env *azure.Environment) (*AZClient, error) {
-	cred, err := auth.NewCredential(cfg)
+	defaultAzureCred, err := azidentity.NewDefaultAzureCredential(nil)
```
Comment (Bryce-Soghigian) [inline pkg/providers/instance/azure_client.go:112]: why change the name cred should suffice

File: pkg/providers/instance/skuclient/skuclient.go [hunk 7]
```diff
@@ -48,11 +50,14 @@ func (sc *skuClient) updateInstance() {
 	sc.mu.RLock()
 	defer sc.mu.RUnlock()
 
-	authorizer, err := auth.NewAuthorizer(sc.cfg, sc.env)
+	// Create a new authorizer for the sku client
+	// TODO (charliedmcb): need to get track 2 support for the skewer API
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
```
Comment (Bryce-Soghigian) [inline pkg/providers/instance/skuclient/skuclient.go:55]: How did you test this?

**Review by Bryce-Soghigian — APPROVED** [review]
Run e2e tests please before merging

File: pkg/auth/config.go (hunk 5 above)
Comment (comtalyst) [reply pkg/auth/config.go:101]: bootstrap variables, mostly

File: pkg/providers/instance/skuclient/skuclient.go (hunk 7 above)
Comment (comtalyst) [reply pkg/providers/instance/skuclient/skuclient.go:55]: It is the code of `auth.NewAuthorizer(sc.cfg, sc.env)`, with changes to use `azidentity.NewDefaultAzureCredential(nil)`.
E2E for each auth method should be sufficient.

**Review by tallaxes — APPROVED** [review]
Overall looks good, noted some potential additional cleanup, but approving in case it is best handled separately.

File: karpenter-values-template.yaml [hunk 8]
```diff
@@ -34,12 +34,7 @@ controller:
       value: ${AZURE_SUBSCRIPTION_ID}
     - name: LOCATION
       value: ${AZURE_LOCATION}
-    # settings for managed workload identity
-    - name: ARM_USE_CREDENTIAL_FROM_ENVIRONMENT
-      value: "true"
-    - name: ARM_USE_MANAGED_IDENTITY_EXTENSION
-      value: "false"
-    - name: ARM_USER_ASSIGNED_IDENTITY_ID
+    - name: ARM_KUBELET_IDENTITY_CLIENT_ID
```
[the remark this answers is not in the document]
Comment (comtalyst) [reply karpenter-values-template.yaml:37]: Not really, let's remove it now.

File: pkg/auth/config.go [hunk 9]
```diff
@@ -61,97 +54,29 @@ type Config struct {
 	ResourceGroup  string `json:"resourceGroup" yaml:"resourceGroup"`
 	VMType         string `json:"vmType" yaml:"vmType"`
```
[the remark this answers is not in the document]
Comment (comtalyst) [reply pkg/auth/config.go:55]: Looks like it, let's clean it up now.

File: pkg/auth/config.go [hunk 10]
```diff
@@ -61,97 +54,29 @@ type Config struct {
...
-	cfg.AADClientID = os.Getenv("ARM_CLIENT_ID")
-	cfg.AADClientSecret = os.Getenv("ARM_CLIENT_SECRET")
-	cfg.VMType = strings.ToLower(os.Getenv("ARM_VM_TYPE"))
-	cfg.AADClientCertPath = os.Getenv("ARM_CLIENT_CERT_PATH")
-	cfg.AADClientCertPassword = os.Getenv("ARM_CLIENT_CERT_PASSWORD")
-	cfg.ClusterName = os.Getenv("AZURE_CLUSTER_NAME")
-	cfg.NodeResourceGroup = os.Getenv("AZURE_NODE_RESOURCE_GROUP")
-}
+	// Managed identity for Kubelet (not to be confused with Azure cloud authorization)
+	KubeletIdentityClientID string `json:"kubeletIdentityClientID" yaml:"kubeletIdentityClientID"`
```
[the remark this answers is not in the document]
Comment (comtalyst) [reply pkg/auth/config.go:58]: No specific reason, but could be clean up along with other options later.

**Review by tallaxes — APPROVED** [review]
